zookeeper-client/zookeeper-client-c/src/mt_adaptor.c

```diff
@@ -502,13 +502,13 @@ int32_t fetch_and_add(volatile int32_t* operand, int incr)
 }
 
 // make sure the static xid is initialized before any threads started
-__attribute__((constructor)) int32_t get_xid()
```

Comment:
why is this being removed? Granted this is from a looong time ago but this is considered MT code and the changelog particularly mentions

>     * get_xid() is not thread-safe (xid initialization race condition 
>       in the multi-threaded mode).

please remove this change from the patch.

Reply:
__attribute__((constructor)) is no longer needed since the following statement was removed:

if (xid == -1) {
    xid = time(0);
}

Having this statement is what made the function not MT safe (that's the racy initialization that the comment talks about).  -- The only way that the MT version of get_xid now manipulates the xid static variable is via an atomic increment.

For the single threaded version of get_xid i stuck with

return xid++ & ~(1<<31);

to keep the behavior of the ST and MT version of get_xid identical, but if you want me to i can change that...

Reply:
To clarify further:

A statement like
   static int xid = 1;

even when within the scope of a function does NOT cause any MT issues, since this places the variable into the initialized data section of the library, hence this gets set when the library gets loaded.

Reply:
gotcha, makes sense. And given we are retaining the warning comment on the method itself we should be OK for future generations... Thanks!